xdsl/parser.py

```diff
@@ -1074,11 +1074,70 @@ def try_parse_type(self) -> Attribute | None:
                             | `{` dialect-attribute-contents+ `}`
                             | [^[]<>(){}\0]+
         """
+        self._synchronize_lexer_and_tokenizer()
         if self.tokenizer.starts_with("!"):
             return self.try_parse_dialect_type()
         else:
             return self.try_parse_builtin_type()
 
+    def parse_optional_attribute(self) -> Attribute | None:
+        """
+        Parse an xDSL attribute, if present.
+        An attribute is either a builtin attribute, which can have various format,
+        or a dialect attribute, with the following format:
+            dialect-attr  ::= `!` attr-name (`<` dialect-attr-contents+ `>`)?
+            attr-name     ::= bare-id
+            dialect-attr-contents ::= `<` dialect-attribute-contents+ `>`
+                            | `(` dialect-attribute-contents+ `)`
+                            | `[` dialect-attribute-contents+ `]`
+                            | `{` dialect-attribute-contents+ `}`
+                            | [^[]<>(){}\0]+
+        """
+        with self.backtracking("attribute"):
```

Comment:
I'd recommend replacing this backtracking as part of this PR. The methods below are all optional, so converting it to be optional by default and raising if `None` should be close to trivial.